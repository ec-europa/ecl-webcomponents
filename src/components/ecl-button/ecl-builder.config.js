const path = require('path');
const pkg = require('./package.json');

const nodeModules = path.resolve(__dirname, '../../../node_modules');
const outputFolder = './build';

// SCSS includePaths
const includePaths = [nodeModules];

const banner = `${pkg.name} - ${pkg.version} Built on ${new Date().toISOString()}`;

module.exports = {
  styles: [
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-button-ec.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-button-ec.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
        minify: true,
      },
    },
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-button-eu.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-button-eu.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
        minify: true,
      },
    },
  ],
};
