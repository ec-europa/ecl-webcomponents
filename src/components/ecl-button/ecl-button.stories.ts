import iconsAllEc from '@ecl/resources-ec-icons/dist/lists/all.json';

const getArgs = () => {
  return {
    label: 'Button label',
    type: 'button',
    variant: 'primary',
    icon: '',
    iconPosition: 'after',
    iconTransform: '',
    hideLabel: false,
    indicator: false,
    indicatorValue: 3,
  };
};

const getArgTypes = () => {
  return {
    color_mode: { table: { disable: true } },
    variant: {
      name: 'button variant',
      type: { name: 'select' },
      options: ['primary', 'secondary', 'ghost', 'call'],
      description: "Button variant"
    },
    type: {
      name: 'button type',
      type: { name: 'select' },
      options: ['button', 'submit'],
      description: "Button type"
    },
    label: {
      name: 'button label',
      type: { name: 'string'},
      description: 'The text in the button',
    },
    hideLabel: {
      name: 'Icon only',
      type: { name: 'boolean'},
      description: 'Hides the label (not for screen-readers)',
      if: { arg: 'icon', truthy: true },
    },
    icon: {
      name: 'icon',
      type: { name: 'select' },
      options: iconsAllEc,
      description: "Add an icon",
    },
    iconPosition: {
      name: 'icon position',
      type: { name: 'select' },
      options: ['before', 'after'],
      description: 'Before or after the button label',
    },
    iconTransform: {
      name: 'icon transform',
      type: { name: 'select' },
      description: 'Button icon transform',
      options: [
        'rotate-90',
        'rotate-180',
        'rotate-270',
        'flip-horizontal',
        'flip-vertical',
      ],
    },
    indicator: {
      name: 'indicator',
      type: { name: 'boolean' },
      description: 'Show an indicator (only visible when label is hidden and icon exists)',
      if: { arg: 'icon', truthy: true },
    },
    indicatorValue: {
      name: 'indicator value',
      type: { name: 'number' },
      description: 'Value of the indicator (only visible when showIndicator is true)',
      if: { arg: 'showIndicator', truthy: false },
    },
  };
};

export default {
  title: 'Components/button',
};

const Template = (args) => {
  return `<ecl-button
    type="${args.type}"
    data-ecl-button-icon
    theme="${args.theme}"
    variant="${args.variant}"
    ${args.hideLabel ? 'hide-label="true"' : ''}s
    ${args.indicator && args.hideLabel ? `indicator indicator-value="${args.indicatorValue}"` : ''}
  >
    ${args.icon && args.iconPosition === 'before' ?
      `<ecl-icon 
        slot="icon-before"
        icon="${args.icon}"
        transform="${args.iconTransform}"
        theme="${args.theme}"
      >
      </ecl-icon>` : ''}
    ${args.label}
    ${args.icon && args.iconPosition === 'after' ?
      `<ecl-icon 
        slot="icon-after"
        icon="${args.icon}"
        theme="${args.theme}"
        transform="${args.iconTransform}"
      >
      </ecl-icon>` : ''}
  </ecl-button>`;
}

export const Button = Template.bind({});
Button.storyName = 'default';
Button.args = getArgs();
Button.argTypes = getArgTypes();
