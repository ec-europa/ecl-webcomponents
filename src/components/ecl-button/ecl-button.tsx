import { Component, Prop, h, Element } from '@stencil/core';

@Component({
  tag: 'ecl-button',
  styleUrls: {
    ec: './build/styles/ecl-button-ec.css',
    eu: './build/styles/ecl-button-eu.css',
  },
  shadow: false,
  scoped: true,
})

export class EclButton {
  @Element() el: HTMLElement;
  @Prop() styleClass: string = '';
  @Prop() type: string = 'primary';
  @Prop() theme: string = 'ec';

  componentDidRender() {
    if (this.el.getElementsByTagName('ecl-icon')[0] && this.el.querySelector('.ecl-icon')) {
      const slot = this.el.getElementsByTagName('ecl-icon')[0].getAttribute('slot');
      this.el.querySelector('.ecl-icon').classList.add('ecl-button__icon');
      if (slot) {
        this.el.querySelector('.ecl-icon').classList.add(`ecl-button__${slot.substring(0, 5) + '-' + slot.substring(5)}`);
      }
    }
  }

  getClass(): string {
    return [
      `ecl-button`,
      `ecl-button--${this.type}`,
      this.styleClass
    ].join(' ');
  }

  render() {
    return (
      <button class={this.getClass()}>
        <span class="ecl-button__container">
          <slot name="icon-before"></slot>
          <span class="ecl-button__label">
            <slot></slot>
          </span>
          <slot name="icon-after"></slot>
        </span>
      </button>
    )
  }
}
