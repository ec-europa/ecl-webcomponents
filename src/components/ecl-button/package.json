{
  "name": "@ecl/web-components-button",
  "author": "European Commission",
  "license": "EUPL-1.2",
  "version": "0.8.0",
  "description": "ECL Button",
  "main": "ecl-button.tsx",
  "private": true,
  "scripts": {
    "build:styles": "cross-env ecl-builder styles",
    "build": "run-s clean build:*",
    "clean": "rimraf build",
    "prepublish": "npm run build"
  },
  "devDependencies": {
    "@ecl/builder": "3.8.4",
    "@ecl/resources-ec-icons": "3.8.4",
    "@ecl/resources-eu-icons": "3.8.4",
    "@ecl/specs-component-button": "3.8.4",
    "@ecl/theme-ec": "3.8.4",
    "@ecl/theme-eu": "3.8.4",
    "@ecl/vanilla-component-button": "3.8.4",
    "@ecl/vanilla-component-icon": "3.8.4",
    "@ecl/vanilla-component-link": "3.8.4",
    "cross-env": "7.0.3"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/planctus/ecl-webcomponents.git"
  },
  "bugs": {
    "url": "https://github.com/planctus/ecl-webcomponents/issues"
  },
  "homepage": "https://github.com/planctus/ecl-webcomponents",
  "keywords": [
    "ecl",
    "europa-component-library",
    "design-system",
    "web-components"
  ]
}
