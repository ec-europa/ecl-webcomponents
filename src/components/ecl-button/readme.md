# ecl-button

<!-- Auto Generated Below -->


## Properties

| Property         | Attribute         | Description | Type      | Default     |
| ---------------- | ----------------- | ----------- | --------- | ----------- |
| `ariaControls`   | `aria-controls`   |             | `string`  | `undefined` |
| `disabled`       | `disabled`        |             | `boolean` | `false`     |
| `hideLabel`      | `hide-label`      |             | `boolean` | `false`     |
| `indicator`      | `indicator`       |             | `boolean` | `false`     |
| `indicatorValue` | `indicator-value` |             | `string`  | `''`        |
| `itemId`         | `item-id`         |             | `string`  | `undefined` |
| `styleClass`     | `style-class`     |             | `string`  | `''`        |
| `theme`          | `theme`           |             | `string`  | `'ec'`      |
| `type`           | `type`            |             | `string`  | `'submit'`  |
| `variant`        | `variant`         |             | `string`  | `'primary'` |


## Dependencies

### Used by

 - [ecl-breadcrumb-item](../ecl-breadcrumb)
 - [ecl-expandable](../ecl-expandable)
 - [ecl-file-translations](../ecl-file)
 - [ecl-gallery](../ecl-gallery)
 - [ecl-menu](../ecl-menu)
 - [ecl-menu-item](../ecl-menu)
 - [ecl-message](../ecl-message)
 - [ecl-modal](../ecl-modal)
 - [ecl-news-ticker](../ecl-news-ticker)
 - [ecl-search-form](../ecl-search-form)
 - [ecl-timeline-item](../ecl-timeline)

### Depends on

- [ecl-indicator](../ecl-indicator)

### Graph
```mermaid
graph TD;
  ecl-button --> ecl-indicator
  ecl-breadcrumb-item --> ecl-button
  ecl-expandable --> ecl-button
  ecl-file-translations --> ecl-button
  ecl-gallery --> ecl-button
  ecl-menu --> ecl-button
  ecl-menu-item --> ecl-button
  ecl-message --> ecl-button
  ecl-modal --> ecl-button
  ecl-news-ticker --> ecl-button
  ecl-search-form --> ecl-button
  ecl-timeline-item --> ecl-button
  style ecl-button fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
