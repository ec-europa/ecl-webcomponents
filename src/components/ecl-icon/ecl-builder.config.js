const path = require('path');
const pkg = require('./package.json');

const nodeModules = path.resolve(__dirname, '../../../node_modules');
const outputFolder = './build';

// SCSS includePaths
const includePaths = [nodeModules];

const banner = `${pkg.name} - ${pkg.version} Built on ${new Date().toISOString()}`;

module.exports = {
  copy: [
    {
      from: [path.resolve(nodeModules, '@ecl/resources-social-media-icons/dist/sprites/icons-social-media.svg')],
      to: path.resolve(outputFolder, 'images/ec'),
      options: { up: true },
    },
    {
      from: [path.resolve(nodeModules, '@ecl/resources-social-media-icons/dist/sprites/icons-social-media.svg')],
      to: path.resolve(outputFolder, 'images/eu'),
      options: { up: true },
    },
    {
      from: [path.resolve(nodeModules, '@ecl/resources-ec-icons/dist/sprites/icons.svg')],
      to: path.resolve(outputFolder, 'images/ec'),
      options: { up: true },
    },
    {
      from: [path.resolve(nodeModules, '@ecl/resources-eu-icons/dist/sprites/icon.svg')],
      to: path.resolve(outputFolder, 'images/eu'),
      options: { up: true },
    },
  ],
  styles: [
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-icon-ec.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-icon-ec.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
        minify: true,
      },
    },
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-icon-eu.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-icon-eu.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
        minify: true,
      },
    },
  ],
};
