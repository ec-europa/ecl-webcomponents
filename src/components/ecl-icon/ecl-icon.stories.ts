import iconPathEc from '@ecl/resources-ec-icons/dist/sprites/icons.svg';
import iconPathEu from '@ecl/resources-eu-icons/dist/sprites/icons.svg';
import iconsAllEc from '@ecl/resources-ec-icons/dist/lists/all.json';

export default {
  title: 'Components/icon',
};

const Template = (args) => `<ecl-icon 
                              icon="${args.icon}"
                              color="${args.color}"
                              size="${args.size}"
                              transform="${args.transform}"
                              path="${args.theme === 'ec' ? iconPathEc : iconPathEu}"
                            ></ecl-icon>`;

export const Icon = Template.bind({});
Icon.storyName = 'default';
Icon.args = {
  theme: 'ec',
  icon: 'close',
  transform: '',
  color: 'default',
  size: '2xl',
};
Icon.argTypes = {
  theme: {
    type: { name: 'select'},
    options: ['ec', 'eu'],
  },
  icon: {
    type: { name: 'select' },
    options: iconsAllEc,
    description: "Add an icon",
  },
  size: {
    type: { name: 'select' },
    options: ['2xs','xs', 's', 'm', 'l', 'xl', '2xl'],
    description: "icon size",
  },
  color: {
    type: { name: 'select' },
    options: ['default', 'primary', 'inverted'],
    description: "icon color",
  },
  transform: {
    type: { name: 'select' },
    description: 'Button icon transform',
    options: [
      'rotate-90',
      'rotate-180',
      'rotate-270',
      'flip-horizontal',
      'flip-vertical',
    ],
  },
};