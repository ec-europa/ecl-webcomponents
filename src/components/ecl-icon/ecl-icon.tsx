import { Component, Prop, h } from '@stencil/core';
import getAssetPath from "../../utils/assetPath";

@Component({
  tag: 'ecl-icon',
  styleUrls: {
    ec: './build/styles/ecl-icon-ec.css',
    eu: './build/styles/ecl-icon-eu.css',
  },
  shadow: false,
  scoped: true,
  assetsDirs: ['build'],
})

export class EclIcon {
  @Prop() styleClass: string = '';
  @Prop() icon: string = '';
  @Prop() size: string = 'xs';
  @Prop() color: string = '';
  @Prop() titleTag: string = '';
  @Prop({ mutable: true }) path: string;
  @Prop() transform: string = '';
  @Prop({ mutable: true }) theme: string;
  @Prop() sprite: string = '';

  componentWillLoad() {
    this.theme = document.documentElement.getAttribute('data-ecl-theme') ?? (this.theme || 'ec');
    
    if (!this.sprite) {
      this.path = getAssetPath(`./build/images/${this.theme}/icons.svg`);
    } else {
      this.path = getAssetPath(`./build/images/${this.theme}/${this.sprite}.svg`);
    }
  }

  getClass(): string {
    const styleClasses = [
      `ecl-icon`,
      `ecl-icon--${this.size}`,
      this.styleClass,
    ];
    if (this.color) {
      styleClasses.push(
        `ecl-icon--${this.color}`
      );
    }
    if (this.transform) {
      styleClasses.push(
        `ecl-icon--${this.transform}`
      );
    }

    return styleClasses.join(' ');
  }

  render() {
    return (
      <svg class={this.getClass()}>
      { this.titleTag ? <title>{this.titleTag}</title> : '' } 
        <use xlinkHref={`${this.path}#${this.icon}`}></use>
      </svg>
    )
  }
}