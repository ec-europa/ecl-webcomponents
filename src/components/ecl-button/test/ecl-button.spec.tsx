import { newSpecPage } from '@stencil/core/testing';
import { EclButton } from '../ecl-button';
import { EclIcon } from '@ecl/web-components-icon/ecl-icon';

describe('ecl-button', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [EclButton, EclIcon],
      html: `<ecl-button
              variant="primary"
              data-ecl-button-icon
              theme="ec"
            >
              <ecl-icon 
                slot="icon-before"
                path="/icons.svg"
                icon="corner-arrow"
                transform="rotate-90"
                style-class="ecl-button__icon ecl-button__icon--before"
              >
              </ecl-icon>
              Button label
            </ecl-button>`,
    });

    expect(page.root).toMatchSnapshot();
  });
});
