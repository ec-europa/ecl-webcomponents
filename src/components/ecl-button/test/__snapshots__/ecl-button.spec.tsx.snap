// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ecl-button renders 1`] = `
<ecl-button class="ecl-button ecl-button--primary" data-ecl-button-icon theme="ec" type="primary">
  <!---->
  <span class="ecl-button__container">
    <ecl-icon icon="corner-arrow" path="/icons.svg" slot="icon-before" style-class="ecl-button__icon ecl-button__icon--before" transform="rotate-90">
      <svg class="ecl-button__icon ecl-button__icon--before ecl-icon ecl-icon--rotate-90 ecl-icon--xs">
        <use xlink:href="/build/images/ec/icons.svg#corner-arrow"></use>
      </svg>
    </ecl-icon>
    <span class="ecl-button__label">
    </span>
  </span>
</ecl-button>
`;
